### PR TITLE
Feature/custom regex pr

### LINKED ISSUE #21: Add support for custom regex
The regex generically derived from the markup is sometimes not right. E.g. for markup `@__id__` and value "@johndoe " it would only match "@j". In such situations, the user should be able to provide a customized regex, such as `/@(\w+)/g`.

### PR DESCRIPTION
**Pull request by nassimbenkirane**
Fixes #21 

What did you change (functionally and technically)?
Now, we should be able to provide a custom regex to the `MarkupInput` component 
so that markup like ':__id__' or '@__display__' can work (see #21)

Example:
```jsx
<MentionsInput
    ...
    markup=':__id__'
    regex={/:(\S+)/g}>
    <Mention
        ...
        trigger='@' />
</MentionsInput>
```
Technically, I took the code in the old custom-regex branch and applied the modifications to the new codebase, adding the regex attribute to a lot of functions in the code. 

* Are there tests for the new code?
No new tests were added, but I updated the tests for `getPositionOfCapturingGroup`

### CODE REVIEW

File: src/Mention.js [hunk 1]
```diff
@@ -6,9 +6,12 @@ const styled = defaultStyle({
   fontWeight: 'inherit',
 })
 
-const Mention = styled(({ display, style }) => (
-  <strong {...style}>{display}</strong>
-))
+class Mention extends React.Component {
```
Comment (frontendphil) [inline src/Mention.js]: Why did you change that from a purely functional component to a regular one? I'd love it to stay functional when possible.

File: src/utils.js [hunk 2]
```diff
@@ -77,10 +79,17 @@ export const getPositionOfCapturingGroup = (markup, parameterName) => {
...
-  if (parameterName === 'display') return sortedIndices.indexOf(indexDisplay)
-  if (parameterName === 'type')
-    return indexType === null ? null : sortedIndices.indexOf(indexType)
+
+  if(regex && countCapturingGroups(regex) === 0) {
+    // custom regex does not use any capturing groups, so use the full match for ID and display
+    return parameterName === "type" ? null : 0;
+  }
+
+  let argOffset = 1; // first argument always is the full match
```
Comment (frontendphil) [inline src/utils.js]: Please use `const` if the value doesn't change.

File: test/utils.spec.js [hunk 3]
```diff
@@ -22,12 +22,12 @@ describe('utils', () => {
 
   describe('#getPositionOfCapturingGroup', () => {
     const testData = {
-      '@[__display__](__id__)': { display: 0, id: 1, type: null },
-      '@[__display__](__type__:__id__)': { display: 0, id: 2, type: 1 },
-      '@(__type__:__id__)': { display: 1, id: 1, type: 0 },
-      '{{__id__#__display__}}': { display: 1, id: 0, type: null },
-      '{{__id__}}': { display: 0, id: 0, type: null },
-      '{{__display__}}': { display: 0, id: 0, type: null },
+      '@[__display__](__id__)': { display: 1, id: 2, type: null },
```
Comment (frontendphil) [inline test/utils.spec.js]: The type of change for the spec makes it look like a breaking API change. Is that the case? I'm also missing a specific test case for mentions that use a `regex` pattern.

File: src/Mention.js (hunk 1 above)
Comment (nassimbenkirane) [reply src/Mention.js]: Hi, thanks for taking the time to look at the code,

I did it some time ago, and my intention was to have it working.
I remember that it was breaking with `styled` trying to add a ref to a non functional component, which resulted in unwanted warnings. I'll double check that

File: src/utils.js (hunk 2 above)
Comment (nassimbenkirane) [reply src/utils.js]: Sure, I took the code that was previously in the custom-regex branch and changed `var`s to `let`s, but this one is definitely a `const` ^^

File: test/utils.spec.js (hunk 3 above)
Comment (nassimbenkirane) [reply test/utils.spec.js]: Is `getPositionOfCapturingGroup` considered a public API ?
I could probably revert this by changing back the `getPlainText` method, but I guess it makes more sense like this. Again, I would be interested in @jfschwarz input since it is his original modifications (from the `custom-regex` branch)

Comment (frontendphil) [reply test/utils.spec.js]: I'd actually need to check this. We do pass props that sound similar to the `render` methods that can be influenced by the user. That's what got me worried. @jfschwarz is on vacation for the next 4 weeks which is why he might not be able to chime in.

File: src/Mention.js (hunk 1 above)
Comment (nassimbenkirane) [reply src/Mention.js]: Upon investigation I had this warning: [#19](https://github.com/jfschwarz/substyle/issues/19), It is fixed in substyle 6.1.1 so I'll update substyle from 6.0.0 to 6.1.1 and revert to using a functional component

Comment (nassimbenkirane) [reply src/Mention.js]: @frontendphil I just committed the changes

File: src/utils.js (hunk 2 above)
Comment (nassimbenkirane) [reply src/utils.js]: 👌

File: test/utils.spec.js (hunk 3 above)
Comment (nassimbenkirane) [reply test/utils.spec.js]: Thanks for your answer :) And enjoy your vacation, @jfschwarz ! Anyway this is not crucial to this PR, so I just reverted the changes, and confirmed that it was still working for me.

The method _is_ exported utils, but _this comment_ tells me it was probably not meant to be publicly available : 
```javascript
/**
 * parameterName: "id", "display", or "type"
 * TODO: This is currently only exported for testing
 */
export const getPositionOfCapturingGroup = (markup, parameterName, regex) => {
```

**Comment by nassimbenkirane** [comment]
Hi @frontendphil, I just added a test for this use case

with
```javascript
const data = [{ id: 'aaaa', display: '@A' }, { id: 'bbbb', display: '@B' }]
```

```jsx
<MentionsInput
    value=":aaaa and :bbbb and :invalidId"
    markup=":__id__"
    regex={/:(\S+)/g}
    displayTransform={id => {
        let mention = data.find(item => item.id === id)
        return mention ? mention.display : `:${id}`
    }}
>
    <Mention trigger="@" data={data} />
</MentionsInput>
```
If we give a value of `:aaaa and :bbbb and :invalidId` to the mentionsInput, we should show `@A and @B and :invalidId` in the textarea

I also did some of the changes that you requested, what do you think ?

**Comment by activescott** [comment]
I also am keen to see this request merged as I think it will eliminate a fork I'm currently using (https://github.com/signavio/react-mentions/compare/master...activescott:markup-regex-hack#diff-2b4ca49d4bb0a774c4d4c1672d7aa781) so I can have mentions like `@email@domain.tld`.

**Comment by activescott** [comment]
I also am keen to see this request merged as it will eliminate a fork I'm currently using (https://github.com/signavio/react-mentions/compare/master...activescott:markup-regex-hack#diff-2b4ca49d4bb0a774c4d4c1672d7aa781) so I can have mentions like `@email@domain.tld`. I have tested it and it seems to work for my needs as well.

**Comment by activescott** [comment]
@frontendphil Are there more changes blocking this from being merged and deployed to NPM? I'm using this fork and it seems to be working for me. I'd love for it to be deployed to npm so we can stop using a fork.

**Comment by frontendphil** [comment]
Published in 2.1.0
